client/service/src/client/client.rs

```diff
@@ -683,11 +683,27 @@ where
 
 		operation.op.insert_aux(aux)?;
 
-		// we only notify when we are already synced to the tip of the chain
+		// We only notify when we are already synced to the tip of the chain
 		// or if this import triggers a re-org
 		if make_notifications || tree_route.is_some() {
 			if finalized {
-				operation.notify_finalized.push(hash);
+				let mut summary = match operation.notify_finalized.take() {
+					Some(summary) => summary,
+					None => FinalizeSummary { finalized: Vec::new(), stale_heads: Vec::new() },
+				};
+				summary.finalized.push(hash);
+				// Add to the stale list all heads that are branching from parent.
+				for head in self.backend.blockchain().leaves()? {
+					if head == parent_hash {
+						continue
+					}
+					let route_from_parent =
+						sp_blockchain::tree_route(self.backend.blockchain(), parent_hash, head)?;
+					if route_from_parent.retracted().is_empty() {
```

Comment:
This is to prevent adding notifications for the heads that we already added in the `apply_finality_with_block_hash`.

At this point (i.e. adding another finalized block on top of the parent and after `apply_finality_with_block_hash`) the only additional stale forks that the `apply_finality_with_block_hash` may have missed are the ones originating from the parent itself without any retraction.

Reply:
All the stale branches and finalized blocks are already included by the `apply_finality_with_block_hash` (above).

At this point, by adding another final block on top of the parent, the only extra stale branches that we may have missed are the ones starting from the parent itself and the condition `route_from_parent.retracted().is_empty()` checks this condition.

What may be missing is to perform this operation only if `parent_exists` is true (otherwise recovering the `tree_route` call will fail).